L58-python-async-api.md

```diff
@@ -0,0 +1,1741 @@
+added to instantiate async `Server` or `Channel` object:
+
+```Python
+import grpc
+server = grpc.aio.server(...)
+channel = grpc.aio.insecure_channel(...)
+channel = grpc.aio.secure_channel(...)
+channel = grpc.aio.intercept_channel(...)
+```
+
+To reduce cognitive burden, this gRFC tries to make the new async API share the same
+parameters as the current API, and keep the usage similar.
+
+### Demo Snippet of Async API
+
+Server side:
+```Python
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def SayHello(self,
+                       request: helloworld_pb2_grpc.HelloRequest,
+                       context: grpc.aio.ServicerContext
+                ) -> helloworld_pb2_grpc.HelloReply:
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message="Hello, %s!" % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+
+    async for response in stub.StreamingHi(...):
+        process(response)
+```
+
+### New Streaming API
+
+Existing streaming API has usability problem that its logic complexes user
+application. For client side, it requires the request iterator to be defined
+above the line of call, and the consumption of responses to be coded below the
+line of invocation. So when the application wants to send request based on
+received response, it would be quite challenging and usually involves
+multi-threading, which is not optimal in Python world.
+
+For server side, the same problem remains for non-trivial streaming servicer
+handlers. The `yield` syntax constraint the message sending should happen in the
+handler itself, otherwise the user needs to code a dedicated iterator class. For
+larger application, it would be a pain to create such iterator class for each
+streaming handler.
+
+euroelessar@ points out several great reasons for the new streaming API:
+* Enable `with` statement for streaming API, which allows clean resource management;
+* Easier to buffer outbound messages when peer pushes back (e.g. `yield` will
+  stop the execution of current coroutine, but `call.write` can be scheduled in
+  event loop);
+* Enable application to have both pending read and pending write in one
+  coroutine, and react on the first finisher (with iterator syntax this will be
+  quite hard to get it right).
+
+So, this gRFC introduces a new pattern of streaming API that reads/writes
+message to peer with explicit call.
+
+#### Snippet For New Streaming API
+
+```Python
+### Client side
+with stub.StreamingHi() as streaming_call:
+  request = echo_pb2.EchoRequest(message="ping")
+  await streaming_call.write(request)
```

Comment:
is this a unary only request or a streaming request?
If streaming, I don't see a way of expressing that client is done sending requests.

See https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core.Api/IClientStreamWriter.cs and https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs for comparison.

Reply:
Yes, the client can tell the server that it's done sending messages. The server will receive a notification (a pending/new read on the server side finishes with "no message"/end of stream), but server can still continue sending responses.  `done_writing()` is ok as a name (I can't speak on how idiomatic that name is in python).  `half_close()` is a bad name because half-close is a HTTP/2 terminology and it's a technical detail that should not be exposed to the end user.

More remarks:
- server of course cannot half close, it can only finish the RPC with a status (which terminates both request and response streams). That also means that the interface for sending  on the client side and server side are going to be different (see the difference in c#) and only client is going to provide `done_writing()`.

Btw, in C# it is customary to add "Async" suffix to names of all async methods, to make it obvious they should be awaited - python doesn't have a style that requires that?

Reply:
`done_writing()` added to the client-side context object.

#### About the "Async" suffix
Python doesn't have that hard requirement, and official modules do not follow that pattern. I have read articles about enforcing the requirement, and I think it may get messier once we introduce that pattern through out our API due to:

* How to name async getters?
* How to name functions that return an `Awaitable` object (like future)?

The function name might get much longer.